runtime/src/accounts.rs

```diff
@@ -267,144 +268,190 @@ impl Accounts {
+        }
+
         // Appends the account_deps at the end of the accounts,
         // this way they can be accessed in a uniform way.
         // At places where only the accounts are needed,
         // the account_deps are truncated using e.g:
         // accounts.iter().take(message.account_keys.len())
         accounts.append(&mut account_deps);
 
-        if validated_fee_payer {
-            let program_indices = message
-                .instructions()
-                .iter()
-                .map(|instruction| {
-                    self.load_executable_accounts(
-                        ancestors,
-                        &mut accounts,
-                        instruction.program_id_index as IndexOfAccount,
-                        error_counters,
-                    )
-                })
-                .collect::<Result<Vec<Vec<IndexOfAccount>>>>()?;
-
-            Ok(LoadedTransaction {
-                accounts,
-                program_indices,
-                rent: tx_rent,
-                rent_debits,
+        let builtins_start_index = accounts.len();
+        let program_indices = message
+            .instructions()
+            .iter()
+            .map(|instruction| {
+                let mut account_indices = Vec::new();
+                let mut program_index = instruction.program_id_index as usize;
+                for _ in 0..5 {
+                    let (program_id, program_account) = accounts
+                        .get(program_index)
+                        .ok_or(TransactionError::ProgramAccountNotFound)?;
+                    let (account_found, account_dep_index) = account_found_and_dep_index
+                        .get(program_index)
+                        .unwrap_or(&(true, None));
+                    if native_loader::check_id(program_id) {
+                        return Ok(account_indices);
+                    }
+                    if !account_found {
+                        error_counters.account_not_found += 1;
+                        return Err(TransactionError::ProgramAccountNotFound);
+                    }
+                    if !program_account.executable() {
+                        error_counters.invalid_program_for_execution += 1;
+                        return Err(TransactionError::InvalidProgramForExecution);
+                    }
+                    account_indices.insert(0, program_index as IndexOfAccount);
+                    if let Some(account_index) = account_dep_index {
+                        account_indices.insert(0, *account_index);
+                    }
+                    let owner_id = program_account.owner();
+                    if native_loader::check_id(owner_id) {
+                        return Ok(account_indices);
+                    }
+                    program_index = if let Some(owner_index) = accounts
+                        .get(builtins_start_index..)
+                        .ok_or(TransactionError::ProgramAccountNotFound)?
```

Comment:
Wondering if it'll always return `None` (mapping to `TransactionError::ProgramAccountNotFound`) here.
https://github.com/solana-labs/solana/pull/29728/files#diff-a6df749724ef3698bf0b90625702f100670b58c585f91da8725f78a94b04196eR391 sets `builtins_start_index` to `accounts.len()`. There is no new entry pushed to `accounts` since `builtins_start_index` was initialized.

Reply:
There is a few lines below: `accounts.push((*owner_id, program_account));`

Reply:
But, the code flow will not reach there, right? Because of this `.ok_or(TransactionError::ProgramAccountNotFound)?` it'll just skip the code block.